src/main/java/org/truffleruby/interop/OutgoingForeignCallNode.java

```diff
@@ -99,8 +103,98 @@ protected int objectId(Object receiver, String name, Object[] args,
             limit = "1")
     protected Object readArrayElement(Object receiver, String name, Object[] args,
             @Cached(value = "name", allowUncached = true) @Shared("name") String cachedName,
+            @CachedLibrary("receiver") InteropLibrary interop,
+            @Cached TranslateInteropExceptionNode translateInteropException,
+            @CachedContext(RubyLanguage.class) RubyContext context,
             @Cached InteropNodes.ReadArrayElementNode readNode) {
-        return readNode.execute(receiver, args[0]);
+        return at(receiver, name, args, cachedName, interop, translateInteropException, context, readNode);
+    }
+
+    @Specialization(
+            guards = {
+                    "name == cachedName",
+                    "cachedName.equals(AT)",
+                    "args.length == 1",
+                    "isBasicInteger(first(args))" },
+            limit = "1")
+    protected Object at(Object receiver, String name, Object[] args,
+            @Cached(value = "name", allowUncached = true) @Shared("name") String cachedName,
+            @CachedLibrary("receiver") InteropLibrary interop,
+            @Cached TranslateInteropExceptionNode translateInteropException,
+            @CachedContext(RubyLanguage.class) RubyContext context,
+            @Cached InteropNodes.ReadArrayElementNode readNode) {
+        try {
+            long size = interop.getArraySize(receiver);
+            long bound = (size != 0) ? size : 0;
+            if (((int) args[0]) < -bound || (int) args[0] >= bound) {
```

Comment:
This cast will crash if `args[0]` is a `long`, a `short` or a `byte`.
Use `LongCastNode` to properly cast it to a `long` no matter what type it is.